### PR TITLE
Safer pack/unpack

### PR DESCRIPTION
**Pull request by jackcmay**
Get rid of the scary-looking unsafe code.  This new code at least triples the number of instructions required but will make auditors a lot more happy

### CODE REVIEW

**Comment by CriesofCarrots** [comment]
Remove these? https://github.com/solana-labs/solana-program-library/blob/74bf8a1f587692a9d14e2afaad2a455de475e308/token/program/src/option.rs#L681

**Comment by jackcmay** [comment]
> Remove these?
> 
> https://github.com/solana-labs/solana-program-library/blob/74bf8a1f587692a9d14e2afaad2a455de475e308/token/program/src/option.rs#L681

They are still being used, why remove?

**Comment by CriesofCarrots** [comment]
> They are still being used, why remove?

Where are `COption::pack` and `COption::unpack_or` still being used, aside from the unit test? This pr removes the usages I'm aware of.

File: token/program/src/pack.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    fn pack_into_slice(&self, dst: &mut [u8]);
+    #[doc(hidden)]
+    fn unpack_from_slice(src: &[u8]) -> Result<Self, ProgramError>;
+
+    /// Borrow `Self` from `input` for the duration of the call to `f`, but first check that `Self`
+    /// is initialized
+    #[inline(never)]
+    fn unpack_mut<U>(
+        input: &mut [u8],
+        f: &mut dyn FnMut(&mut Self) -> Result<U, ProgramError>,
```
Comment (garious) [inline token/program/src/pack.rs:29]: Why the trait object here and not `&mut F where F: FnMut(&mut Self) -> Result<U, ProgramError>`?

**Comment by garious** [comment]
A shame this doesn't compile down to something just as fast. A lot of that code looks equivalent to me. Oh well - definitely easier to audit this way.

File: token/program/src/pack.rs (hunk 1 above)
Comment (jackcmay) [reply token/program/src/pack.rs]: That'll work too, updated

**Comment by jackcmay** [comment]
@garious Part of it I'm sure is the copy into and out of that's happening in these changes that was not happening before.

**Pull request merged**